### PR TITLE
Fix schema link to return errors for unknown queries

### PR DESCRIPTION
**Pull request by amannn**
### Checklist:

- ~If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes~
- [x] Make sure all of the significant new logic is covered by tests

Please see the inline comments. I also fixed some tests that I think didn't quite do what they looked like.

### CODE REVIEW

File: src/link/schema/__tests__/schemaLink.ts [hunk 1]
```diff
@@ -56,7 +54,9 @@ describe('SchemaLink', () => {
     });
     observable.subscribe({
       next,
-      error: error => expect(false),
```
Comment (amannn) [inline src/link/schema/__tests__/schemaLink.ts:59]: `expect(false)` doesn't seem to do what it looks like. I've replaced this with throwing an error that would crash the test.

File: src/link/schema/__tests__/schemaLink.ts [hunk 2]
```diff
@@ -65,23 +65,26 @@ describe('SchemaLink', () => {
...
-      result => expect(false),
-      error => {
-        expect(error).toEqual(mockError.throws);
-        done();
-      },
-      () => {
-        expect(false);
-        done();
-      },
-    );
```
Comment (amannn) [inline src/link/schema/__tests__/schemaLink.ts:84]: Also this test didn't really assert for anything.

File: src/link/schema/__tests__/schemaLink.ts [hunk 3]
```diff
@@ -101,7 +104,9 @@ describe('SchemaLink', () => {
     });
     observable.subscribe(
       next,
-      error => expect(false),
```
Comment (amannn) [inline src/link/schema/__tests__/schemaLink.ts:104]: Same here as mentioned above.

File: src/link/schema/__tests__/schemaLink.ts [hunk 4]
```diff
@@ -192,4 +197,22 @@ describe('SchemaLink', () => {
...
+        }
+      `
+    });
+    observable.subscribe(result => {
+      expect(result.errors).toBeTruthy()
+      expect(result.errors!.length).toBe(1)
+      expect(result.errors![0].message).toMatch(/Cannot query field "unknown"/)
+      done();
+    });
+  });
```
Comment (amannn) [inline src/link/schema/__tests__/schemaLink.ts:217]: This is the new test I've added that now passes.

File: src/link/schema/index.ts [hunk 5]
```diff
@@ -48,14 +49,21 @@ export class SchemaLink extends ApolloLink {
...
+        }
+
+        return execute(
+          this.schema,
+          operation.query,
+          this.rootValue,
+          context,
+          operation.variables,
+          operation.operationName,
+        )
```
Comment (amannn) [inline src/link/schema/index.ts:65]: Usually when you use `graphql` to execute a query, the query is validated against the schema. When you use the lower level `execute`, this validation is skipped and an empty object would be returned. The real error is swallowed.

We could theoretically use `graphql` which includes the validation, but then we have to `print` the schema again, since the function doesn't accept a parsed `DocumentNode`. Internally it then parses the query again.

Therefore I think the better solution is add the necessary validation here instead and continue to use `execute`.

See https://github.com/graphql/graphql-js/blob/3aad20bf1cccda59ccb8d855584097fcf7348fef/src/graphql.js#L189-L193

File: src/link/schema/__tests__/schemaLink.ts (hunk 1 above)
Comment (benjamn) [reply src/link/schema/__tests__/schemaLink.ts:59]: Thanks for noticing this and for going out of your way to improve it!

File: src/link/schema/index.ts (hunk 5 above)
Comment (benjamn) [reply src/link/schema/index.ts:65]: I agree with this reasoning!

Including the `graphql/validation/validate` dependency tree will increase bundle sizes, but only for apps using `@apollo/client/link/schema`—and of course one good solution is to move schema/execution/validation-related logic to a GraphQL server, instead of using this link.

**Review by benjamn — APPROVED** [review]
@amannn Thanks for noticing this problem and fixing it, with tests!

File: src/link/schema/index.ts (hunk 5 above)
Comment (amannn) [reply src/link/schema/index.ts:65]: Thanks for the review!

> one good solution is to move schema/execution/validation-related logic to a GraphQL server, instead of using this link.

Yep, I agree! At least for me the use cases for the schema link so far were to use it for SSR when the UI server also provides the GraphQL endpoint or for providing mock data in tests. In both cases bundle size shouldn't be an issue IMO 🙂.

If you're using the schema link client side, you're including the whole schema with all resolvers, which I guess already has quite a bundle size overhead …

Comment (amannn) [reply src/link/schema/index.ts:65]: @benjamn Also thanks for adding the changelog entry!

I just saw this wording:

> The schema link package (@apollo/client/link/schema) will now validate incoming queries against its client-side schema, and return errors as a GraphQL server would.

I'm wondering if the "client-side schema" and "server" part could be a bit misleading? Using the schema link implies that you're providing a local schema, regardless of if it's server or client side, right?

Maybe this could be an alternative?

> The schema link package (`@apollo/client/link/schema`) will now validate incoming queries against the schema before execution and therefore now has the same error handling capabilities as what a developer might be familiar with from a GraphQL endpoint or using `graphql` directly.

Not sure if it's too wordy? I'm also not a native English speaker, so there's likely something to be improved 🙂.

Comment (benjamn) [reply src/link/schema/index.ts:65]: @amannn Improved the wording a bit in https://github.com/apollographql/apollo-client/pull/7015/commits/e27be1e31883c4a711206b1e7d1f179be3bbcc17.

**Comment by benjamn** [comment]
@amannn (+ anyone else who comes across this PR) I got a bit nervous about this being a breaking change for a minor release (AC3.3), so I made the validation opt-in in https://github.com/apollographql/apollo-client/pull/7015/commits/e27be1e31883c4a711206b1e7d1f179be3bbcc17.

**Comment by amannn** [comment]
@benjamn I understand, thanks for the hint!

Do you have a use case in mind where this would be breaking? The previous behaviour was that an empty object was returned – regardless of what you query. In the apps where I noticed this behaviour, this broke the app, but in more subtle ways (runtime error in components) instead of warning that the query is invalid as soon as it is fired.
